asserts/asserts.go

```diff
@@ -218,11 +220,173 @@ func Decode(serializedAssertion []byte) (Assertion, error) {
+		d.err = err
+	}
+	return buf, d.err
+}
+
+func (d *Decoder) readExact(size int) ([]byte, error) {
+	buf, err := d.peek(size)
+	d.b.Discard(len(buf))
+	if len(buf) == size {
+		return buf, nil
+	}
+	if err == io.EOF {
+		return buf, io.ErrUnexpectedEOF
+	}
+	return buf, err
+}
+
+func (d *Decoder) readUntil(delim []byte, maxSize int) ([]byte, error) {
+	last := 0
+	size := d.initialBufSize
+	for {
+		buf, err := d.peek(size)
+		if i := bytes.Index(buf[last:], delim); i >= 0 {
+			d.b.Discard(last + i + len(delim))
+			return buf[:last+i+len(delim)], nil
+		}
+		// report errors only once we have consumed what is buffered
+		if err != nil && len(buf) == d.b.Buffered() {
+			d.b.Discard(len(buf))
+			return buf, err
+		}
+		last = size - len(delim) + 1
+		size *= 2
+		if size > maxSize {
+			return nil, fmt.Errorf("max size exceeded")
+		}
+	}
+}
+
+// Decode parses the next assertion from the stream.
+// It returns the error io.EOF at the end of a well-formed stream.
+func (d *Decoder) Decode() (Assertion, error) {
+	// read the headers and the nlnl separator after them
+	headAndSep, err := d.readUntil(nlnl, d.maxHeadSize)
+	if err != nil {
+		if err == io.EOF {
+			if len(headAndSep) != 0 {
+				return nil, io.ErrUnexpectedEOF
+			}
+			return nil, io.EOF
+		}
+		return nil, fmt.Errorf("error reading assertion head: %v", err)
+	}
+
+	contentBuf := bytes.NewBuffer(make([]byte, 0, len(headAndSep)))
+	contentBuf.Write(headAndSep)
+	head := contentBuf.Bytes()[:len(headAndSep)-len(nlnl)]
+
+	headers, err := parseHeaders(head)
+	if err != nil {
+		return nil, fmt.Errorf("parsing assertion headers: %v", err)
+	}
+
+	length, err := checkInteger(headers, "body-length", 0)
+	if err != nil {
+		return nil, fmt.Errorf("assertion: %v", err)
+	}
+	if length > d.maxBodySize {
+		return nil, fmt.Errorf("assertion body length %d exceeds maximum body size", length)
+	}
+
+	var body []byte
+	var sig []byte
+	if length > 0 {
+		// read the body if length != 0
+		body, err = d.readExact(length)
+		if err != nil {
+			return nil, err
+		}
+		bodyStart := contentBuf.Len()
+		contentBuf.Write(body)
+		body = contentBuf.Bytes()[bodyStart:]
+	}
+
+	// try to read the end of body a.k.a content/signature separator
+	endOfBody, err := d.readUntil(nlnl, d.maxSigSize)
+	if err != nil && err != io.EOF {
+		return nil, fmt.Errorf("error reading assertion trailer: %v", err)
+	}
+
+	var content []byte
+	if bytes.Equal(endOfBody, nlnl) {
+		// we got the nlnl content/signature separator, read the signature now and the assertion/assertion nlnl separation
+		sig, err = d.readUntil(nlnl, d.maxSigSize)
+		if err != nil && err != io.EOF {
+			return nil, fmt.Errorf("error reading assertion signature: %v", err)
+		}
+		content = contentBuf.Bytes()
+	} else {
+		// we got the signature directly which is a ok format only if body length == 0
+		if length > 0 {
+			return nil, fmt.Errorf("missing content/signature separator")
+		}
+		sig = endOfBody
+		content = head
```

Comment:
The logic here has some curious consequences. Because we continue to write into contentBuf after we've collected bytes out of it, it means we can have pointers to different backing arrays, despite having the content all at once in memory at the end.

Although there's no actual bug here yet other than more memory being pinned, it's hard to predict how much that function will change after we stop looking. I'd suggest computing the arguments to Assemble all at the end, to avoid the curious behavior.